libs/client-api/src/native/retry.rs

```diff
@@ -1,68 +1,71 @@
-use crate::notify::ClientToken;
 use crate::ws::{ConnectState, ConnectStateNotify, CurrentAddr, StateNotify, WSError};
-use app_error::gotrue::GoTrueError;
-use gotrue::grant::{Grant, RefreshTokenGrant};
-use parking_lot::RwLock;
 use std::future::Future;
 use std::pin::Pin;
-use std::sync::{Arc, Weak};
+use std::sync::Weak;
 use std::time::Duration;
 use tokio_retry::strategy::FixedInterval;
 use tokio_retry::{Action, Condition, RetryIf};
 use tracing::{debug, info};
 use websocket::{connect_async, WebSocketStream};
 
-pub(crate) struct RefreshTokenAction {
-  token: Arc<RwLock<ClientToken>>,
-  gotrue_client: Arc<gotrue::api::Client>,
-}
-
-impl RefreshTokenAction {
-  pub fn new(token: Arc<RwLock<ClientToken>>, gotrue_client: gotrue::api::Client) -> Self {
-    Self {
-      token,
-      gotrue_client: Arc::new(gotrue_client),
-    }
-  }
-}
-
-impl Action for RefreshTokenAction {
-  type Future = Pin<Box<dyn Future<Output = Result<Self::Item, Self::Error>> + Send + Sync>>;
-  type Item = ();
-  type Error = GoTrueError;
-
-  fn run(&mut self) -> Self::Future {
-    let weak_token = Arc::downgrade(&self.token);
-    let weak_gotrue_client = Arc::downgrade(&self.gotrue_client);
-    Box::pin(async move {
-      if let (Some(token), Some(gotrue_client)) =
-        (weak_token.upgrade(), weak_gotrue_client.upgrade())
-      {
-        let refresh_token = token
-          .read()
-          .as_ref()
-          .ok_or(GoTrueError::NotLoggedIn(
-            "fail to refresh user token".to_owned(),
-          ))?
-          .refresh_token
-          .as_str()
-          .to_owned();
-        let access_token_resp = gotrue_client
-          .token(&Grant::RefreshToken(RefreshTokenGrant { refresh_token }))
-          .await?;
-        token.write().set(access_token_resp);
-      }
-      Ok(())
-    })
-  }
-}
-
-pub(crate) struct RefreshTokenRetryCondition;
-impl Condition<GoTrueError> for RefreshTokenRetryCondition {
-  fn should_retry(&mut self, error: &GoTrueError) -> bool {
-    error.is_network_error()
-  }
-}
+// pub(crate) struct RefreshTokenAction {
+//   token: Arc<RwLock<ClientToken>>,
+//   gotrue_client: Arc<gotrue::api::Client>,
+// }
+//
+// impl RefreshTokenAction {
+//   pub fn new(token: Arc<RwLock<ClientToken>>, gotrue_client: gotrue::api::Client) -> Self {
+//     Self {
+//       token,
+//       gotrue_client: Arc::new(gotrue_client),
+//     }
+//   }
+// }
+//
+// impl Action for RefreshTokenAction {
```

Comment:
How to auto retry if refresh fail?

Reply:
does current implementation and keeps waiting forever? there's issues integrating with ClientWithMiddleware. Not sure if there's another way.

Reply:
Nope. That depends on the strategy you use.

<img width="892" alt="image" src="https://github.com/AppFlowy-IO/AppFlowy-Cloud/assets/86001920/f518303e-057e-4bfb-b8f8-4b51ee28ae27">